### ISSUE TITLE
PDF randomly loses beginning pages while splitting into images

### ISSUE THREAD

**Issue by laritakr** [scientist-softserv/iiif_print#233]
# Story

Put on your detective hat... this one is a puzzle!

This issue was first noticed with works deposited 4/25/2023 on Adv prod. It has appeared in both Adv prod and BL staging. Adv production is on the earlier of the two versions of iiif_print at hash 53ecfeb4ef90ba8dcdbe84f0126598b6f9b33279 so the issue began prior to this commit.

Refs:
- BL ticket: https://github.com/scientist-softserv/britishlibrary/issues/401
- Adv slack thread: https://assaydepot.slack.com/archives/C0313NJV9PE/p1682610040930439

# Acceptance Criteria

- [ ] Identify and remediate incomplete PDF splitting issue
- [ ] update gem in affected repos

# Screenshots / Video

<details>
<summary></summary>

</details>

# Testing Instructions and Sample Files

-

# Notes

The work that BL noted in their ticket has some oddities that should be noted. 

- The child works which were appearing on the homepage (and still appear in the catalog search) are actually linked as child works on the parent, so did they not get indexed as children? If they had that term indexed, they would not be appearing.
- Child work relationships are only formed if all of the expected children have been created. This implies that the splitting failed to form all of the image files and somehow started on page 39 with nothing failing... the number of images created matches the number of works created.

The earliest version of IiifPrint being used where this issue appeared is on hash 53ecfeb4ef90ba8dcdbe84f0126598b6f9b33279
Earliest known imports exhibiting the issue happened 4/25/2023

This [iiif_print commit](https://github.com/scientist-softserv/iiif_print/pull/202) looks particularly suspicious given the reported behavior and may be a place to start looking. It incorporated a large refactor of PDF splitting code.

Another less likely commit [is this one](https://github.com/scientist-softserv/iiif_print/pull/208), as it optimized the identify command.

A third commit that could be in play would be this one, as it [adjusted how we build the name of the work's title](https://github.com/scientist-softserv/iiif_print/pull/211).

**Comment by laritakr** [comment]
## Working notes

After more review, I found that the child works all exist. This is a case where a variety of things could be wrong with the data:

- is_page_of_ssim is sometimes empty in some of the children
- is_child_bsi is sometimes empty in some of the children
      Both of the above come from the `child_indexer`, but they are not always both in error. We can have one but not the other present (for either of them)

- some child works are missing from `ordered_members`, though they all existed in `member_ids`
From Samvera Slack:
      ```
      Depending on the context I believe it is possible that one of those might be getting pulled from solr while the other from fedora.  So you could try reindexing all of the members and see if that helps.
      ```

- IiifPrint::PendingRelationship records remained in the database, indicating the job did not complete normally (I could not find an errors in Sentry). 

So in short, it appears that an error in the actor stack is causing the job to not complete normally.

**Comment by laritakr** [comment]
It looks like the relationships in some cases are not fully formed. 

Relationships are created using `Hyrax::CurationConcern.actor.update(env)` on the parent work to assign `work_members_attributes` which does not appear to have failed. `AttachMembersActor` handles this attribute. 

The ones in error have `member_of_works` but not `in_works`. 
`member_of_works` comes from Hydra::Works::WorkBehavior using `in_objects` & `work?`
`in_works` comes from Hydra::Works::WorkBehavior using `ordered_by` from ActiveFedora::Aggregation::BaseExtension where parent class includes Hydra::Works::WorkBehavior

So the `ordered_members` relationship got missed while the members relationship themselves were added. 

There were some works which indexing appeared to show they were correct, but reindexing lost the `is_page_of_ssim` term which relies on `in_works`, so apparently the ordered_members relationship got removed AFTER originally being there.

**Comment by laritakr** [comment]
See https://github.com/samvera-labs/bulkrax/issues/804 for an explanation of the issue. 

**Solution:** Need to rewrite the relationships job matching the relevant portions of the rewrite of the Bulkrax relationships job, and adding in a lock on the parent work.
